api/xds_protocol.rst

```diff
@@ -155,45 +231,109 @@ the resources in the :ref:`DiscoveryResponse <envoy_api_msg_DiscoveryResponse>`
+
+For `Listener` and `Cluster` resource types, because each response needs to include all resources requested by the client,
+if a client requests a resource that does not exist, it can immediately tell this from the response.
+
+However, for other resource types, because each resource can be sent in its own response, there is no way to know from the
+next response whether the newly requested resource exists, because the next response could be an unrelated update for another
+resource that had already been subscribed to previously. As a result, clients are expected to use a timeout (recommended
+duration is 15 seconds) after sending a request for a new resource, after which they will consider the requested resource to
+not exist if they have not received the resource.
+
+Note that even if a requested resource does not exist at the moment when the client requests it, that resource could be
+created at any time. Management servers must remember the set of resources being requested by the client, and if one of
+those resources springs into existence later, the server must send an update to the client informing it of the new resource.
+Clients that initially see a resource that does not exist must be prepared for the resource to be created at any time.
+
+Unsubscribing From Resources
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+In the incremental API variants, resources can be unsubscribed to via the :ref:`resource_names_unsubscribe
+<envoy_api_field_DeltaDiscoveryRequest.resource_names_unsubscribe>` field.
+
+In the SotW API variants, unsubscriptions are more complicated. For any request where the :ref:`resource_names
+<envoy_api_field_DiscoveryRequest.resource_names>` field is specified, a resource may be unsubscribed to by sending a new
+request without the resource name that the client wishes to unsubscribe from. For example, if the client had previously been
+subscribed to resources A and B but then sends a new request where the resource_names field contains only resource A, that will
+unsubscribe from B.
+
+For `Listener` and `Cluster` resource types, if resource_names is not specified in the request, then the server is responsible
+for selecting the set of resources needed for the client, so there is no mechanism for unsubscribing to one particular resource.
+If resource_names *is* specified, then resources can be unsubscribed from as long as there is still at least one other
+resource still being subscribed to, but there is no mechanism to unsubscribe from the last resource on the stream, because
+sending a request with an empty resource_names list would trigger the server's wildcard behavior. If using the
+non-aggregated SotW API, the client can simply terminate the stream in this case, but if using ADS, the client may need to
+continue subscribing to the last resource even though it does not want to, because terminating the stream would cause
+disruption for other resource types that may need to continue receiving updates.
```

Comment:
It would be valid IMHO to express here that we suggest incremental is really the right thing for this kind of client.

Reply:
Does the incremental API actually solve this problem?  This is basically the issue that @sanjaypujare and I were discussing above.  My assumption was that in the incremental API, if the client unsubscribes to all Listener or Cluster resources, that would revert to the wildcard behavior, where the client did not specify any resources to subscribe to in the first place.  Is that not the case?

Hmmm... Actually, another option would be to say that if the client had previously specified at least one resource name on the Listener or Cluster stream (or sub-stream, in the case of ADS), then if it removes the last resource name, that means to unsubscribe to all instead of revert to the wildcard behavior.  In other words, the wildcard behavior would happen only if the initial request on the stream does not contain any resource names.  I think this approach would work for both the incremental and SotW variants.  WDYT?

Reply:
I'm good with this approach; I'd like to simplify this paragraph, it has too jankiness around stream manipulation.

Reply:
I've rewritten the paragraph to describe the new proposal.  It seems easier to understand now.